src/main/scala/com/github/gvolpe/fs2/Counter.scala

```diff
@@ -22,30 +17,31 @@ object CounterApp extends Counter[IO]
   * #2 >> 3
   * #3 >> 3
   * */
-class Counter[F[_] : Effect] extends StreamApp[F] {
-
-  override def stream(args: List[String], requestShutdown: F[Unit]): fs2.Stream[F, ExitCode] =
-    Scheduler(corePoolSize = 10).flatMap { implicit S =>
-      for {
-        ref <- Stream.eval(async.refOf[F, Int](0))
-        w1  = new Worker[F](1, ref)
-        w2  = new Worker[F](2, ref)
-        w3  = new Worker[F](3, ref)
-        ec  <- Stream(w1.start, w2.start, w3.start).join(3).drain ++ Stream.emit(ExitCode.Success)
-      } yield ec
-    }
+object CounterApp extends IOApp {
+  import cats.syntax.all._
+
+  def stream[F[_]: ConcurrentEffect]: Stream[F, Unit] =
+    for {
+      ref <- Stream.eval(Ref.of[F, Int](0))
+      w1 = new Worker[F](1, ref)
+      w2 = new Worker[F](2, ref)
+      w3 = new Worker[F](3, ref)
+      _ <- Stream(w1.start, w2.start, w3.start).parJoin(3)
+    } yield ()
 
+  override def run(args: List[String]): IO[ExitCode] =
+    stream[IO].compile.drain.as(ExitCode.Success)
 }
 
-class Worker[F[_]](number: Int, ref: Ref[F, Int])
-                  (implicit F: Effect[F]) {
+class Worker[F[_]](number: Int, ref: Ref[F, Int])(implicit F: Effect[F]) {
```

Comment:
I think the constraint can be relaxed here, don't think we need `Effect`. Probably `Sync` will suffice.

Reply:
done